protocols/gossipsub/src/protocol.rs

```diff
@@ -183,18 +182,24 @@ where
 
 /* Gossip codec for the framing */
 
-pub struct GossipsubCodec {
+pub struct GossipsubCodec<In> {
     /// Codec to encode/decode the Unsigned varint length prefix of the frames.
     length_codec: codec::UviBytes,
     /// Determines the level of validation performed on incoming messages.
     validation_mode: ValidationMode,
+    codec: prost_codec::Codec<In>,
 }
 
-impl GossipsubCodec {
-    pub fn new(length_codec: codec::UviBytes, validation_mode: ValidationMode) -> Self {
+impl<In> GossipsubCodec<In> {
+    pub fn new(
+        length_codec: codec::UviBytes,
+        validation_mode: ValidationMode,
+    ) -> GossipsubCodec<In> {
+        let mut codec = prost_codec::Codec::new(0);
         GossipsubCodec {
```

Comment:
@thomaseizinger (When you've got time, no rush) So, this is where things went off the rails for me earlier. I believe this is how that codec field would be initialized.

But... This seems to cause a cascade of updates to everywhere `GossipsubCodec` is referenced. When I add the `<In>` to other spots (the impl for Encoder for example) it complains that it can not find the type for `In` in the scope. One thing I find weird is that it let me get this much code in without specifying a type for `In`, but yet other references that are "further out" are now being flagged as needing a type.

I found [this SO answer](https://stackoverflow.com/a/30019078/277163) that looks like it is what I need, but when I tried implementing it, even more things started showing up as build errors. (my expectation is the compiler would handle this for me, but clearly this isn't the case.)

So my question: is this on the right path for what you were suggesting? I can keep going on this, but just the sheer volume of problems I saw come up made me pause.

Reply:
The type parameter of `prost_codec::Codec` is the message that we are deserializing. In the context of `GossipsubCodec`, that is always the same protobuf so instead forwarding the `In` type parameter to the callers of `GossipsubCodec`, you can set it to a type within the field definition of `codec`, more specifically, you can set it to `rpc_proto::Rpc`:

```rust
codec: prost_codec::Codec<rpc_proto::Rpc>,
```

Reply:
That should get rid of most compile errors :)

Try to zoom out a bit in what we are doing here (I know this can be hard when you are just learning the language):

- `prost_codec::Codec` is an abstraction that combines a `codec::UviBytes` together with a protobuf deserializer: It is a TLV-kind of encoding: We read the length from the socket, then we read the number of bytes described in the length and then we deserialize the message into a typed data structure.
- We are doing the exact same thing already in gossipsub, just all "inlined", i.e. without an abstraction.
- The goal of the issue you've taken on is to remove this code duplication and use the `prost_codec::Codec` abstraction instead of implementing it all manually in here!

Reply:
Most notable, the API of `GossipsubCodec` should remain completely untouched as this is a pure refactoring. Start with that as a constraint and work from there :)

Reply:
Awesome advice @thomaseizinger, thank you so much! I've got that in place and tests passing locally and it doesn't look nearly as rough as it did before. :sweat_smile: I'm going to revisit in the morning to double check it, etc.

> the API of GossipsubCodec should remain completely untouched as this is a pure refactoring

Thank you for that, I was wondering that but completely forgot to ask.

Reply:
Sweet, looking forward to you pushing your changes!